src/server/frontend_wayland/window_wl_surface_role.h

```diff
@@ -76,7 +80,7 @@ class WindowWlSurfaceRole : public WlSurfaceRole
     void set_title(std::string const& title);
     void initiate_interactive_move();
     void initiate_interactive_resize(MirResizeEdge edge);
-    void set_parent(optional_value<SurfaceId> parent_id);
+    void set_parent(std::experimental::optional<std::weak_ptr<scene::Surface>> const& parent);
```

Comment:
Not really related to this review but...

Seeing the spread of the `std::experimental::optional<std::weak_ptr<Type>>` idiom around the code suggests we're missing an abstraction (and a good name for it).

Reply:
I'm not sure quite what abstraction you're thinking of, but in this particular case I determined it would be better to pass around an optional shared pointer.

Reply:
The abstraction implemented as "an optional shared pointer" but with a meaningful name: `std::experimental::optional<std::weak_ptr<scene::Surface>> const&` is a bit of a mouthful.